### PR TITLE
ElementDefinition.constrainType

### PR DESCRIPTION
**Pull request by cmoesel**
This PR adds the implementation of `ElementDefinition.constrainType`.  See the inline documentation and tests for information about what it supports and how to use it.

It also updates the grammar and importer to better support specifying references on the right-hand-side of the `only` rule.

### CODE REVIEW

File: src/fhirtypes/ElementDefinition.ts [hunk 1]
```diff
@@ -238,6 +240,290 @@ export class ElementDefinition {
...
+   * - attempting to add a profile that doesn't match any of the existing types
+   * - attempting to add a base reference that wasn't already a reference
+   * - attempting to add a reference to a profile that doesn't match any of the existing references
+   * - specifying a target that does not match any of the existing type choices
+   * - specifying types or a target whose definition cannot be found
+   * @see {@link http://hl7.org/fhir/R4/elementdefinition-definitions.html#ElementDefinition.type}
+   * @param {{ type: string; isReference?: boolean }[]} types - the set of constrained types,
+   *   identified by id/type/url strings and an optional reference flag (defaults false)
+   * @param {ResolveFn} resolve - a function that can resolve a type to a StructureDefinition instance
+   * @param {string} [target] - a specific target type to constrain.  If supplied, will attempt to
```
Comment (kjmahalingam) [inline src/fhirtypes/ElementDefinition.ts]: Just a nitpick, but is there a reason that `target` is in brackets here? Is that some notation I don't understand, or is it a mistake?

File: src/fhirtypes/ElementDefinition.ts [hunk 2]
```diff
@@ -238,6 +240,290 @@ export class ElementDefinition {
...
+
+  /**
+   * Given an input type (the constraint) and a set of target types (the things to potentially
+   * constrain), find the match and return information about it.
+   * @param {{ type: string; isReference?: boolean }} type - the constrained types, identified by
+   *   id/type/url string and an optional reference flag (defaults false)
+   * @param {ElementDefinitionType[]} targetTypes - the element types that the constrained type
+   *   can be potentially applied to
+   * @param {ResolveFn} resolve - a function that can resolve a type to a StructureDefinition instance
+   * @param {string} [target] - a specific target type to constrain.  If supplied, will attempt to
```
Comment (kjmahalingam) [inline src/fhirtypes/ElementDefinition.ts]: Same question as before.

File: src/fhirtypes/ElementDefinition.ts [hunk 3]
```diff
@@ -238,6 +240,290 @@ export class ElementDefinition {
...
+   * @throws {InvalidTypeError} when the type doesn't match any of the targetTypes
+   */
+  private findTypeMatch(
+    type: { type: string; isReference?: boolean },
+    targetTypes: ElementDefinitionType[],
+    resolve: ResolveFn
+  ): ElementTypeMatchInfo {
+    let matchedType: ElementDefinitionType;
+
+    // Get the lineage (type hierarchy) so we can walk up it when attempting to match
```
Comment (kjmahalingam) [inline src/fhirtypes/ElementDefinition.ts]: Lineage is a very nice term for this. I like it.

File: src/fhirtypes/ElementDefinition.ts [hunk 4]
```diff
@@ -238,6 +240,290 @@ export class ElementDefinition {
...
+
+    return results;
+  }
+
+  /**
+   * Given a new ElementTypeDefinition (based on the existing one), will apply the matching
+   * profiles and targetProfiles as appropriate.  If a targetType was specified, will filter out
+   * the other profiles or targetProfiles.
+   * @param {ElementDefinitionType} newType - the new type to apply the profiles/targetProfiles to
+   * @param {ElementDefinitionType} [targetType] - the (potentially null) target type for the
```
Comment (kjmahalingam) [inline src/fhirtypes/ElementDefinition.ts]: I'm beginning to guess that this is a convention I don't understand. Perhaps being the convention for optional arguments?

File: src/fhirtypes/ElementDefinition.ts [hunk 5]
```diff
@@ -498,8 +784,13 @@ export class ElementDefinition {
...
    */
   unfold(resolve: ResolveFn = () => undefined): ElementDefinition[] {
-    if (this.type.length === 1) {
-      const def = resolve(this.type[0]);
+    if (
+      this.type.length === 1 &&
+      (this.type[0].profile == null || this.type[0].profile.length <= 1)
+    ) {
+      // If it has a profile, use that, otherwise use the code
+      const type = this.type[0].profile?.[0] ?? this.type[0].code;
```
Comment (kjmahalingam) [inline src/fhirtypes/ElementDefinition.ts]: This is the most beautiful line of code I've ever seen.

File: src/fhirtypes/ElementDefinition.ts (hunk 1 above)
Comment (cmoesel) [reply src/fhirtypes/ElementDefinition.ts]: In JSDoc, you can use brackets to indicate that an argument is optional or can be null.

File: src/fhirtypes/ElementDefinition.ts (hunk 2 above)
Comment (cmoesel) [reply src/fhirtypes/ElementDefinition.ts]: In JSDoc, you can use brackets to indicate that an argument is optional or can be null.

File: src/fhirtypes/ElementDefinition.ts (hunk 3 above)
Comment (cmoesel) [reply src/fhirtypes/ElementDefinition.ts]: Thanks.  I felt a bit like I was a historian studying royalty, but it was fun nonetheless.

File: src/fhirtypes/ElementDefinition.ts (hunk 4 above)
Comment (cmoesel) [reply src/fhirtypes/ElementDefinition.ts]: 🥇

File: src/fhirtypes/ElementDefinition.ts (hunk 5 above)
Comment (cmoesel) [reply src/fhirtypes/ElementDefinition.ts]: There are a lot of `?`s, but it's definitely better than:
```js
let type;
if (this.type[0].profile && this.type[0].profile.length > 0) {
  type = this.type[0].profile[0];
} else {
  type = this.type[0].code;
}
```

**Pull request merged**